app/fee_distribution.go

```diff
@@ -48,6 +49,66 @@ func (vac *ValAddrCache) GetAccAddr(ctx sdk.Context, consAddr sdk.ConsAddress) s
 	return accAddr
 }
 
+func (vac *ValAddrCache) SetDistributionAddr(consAddr sdk.ConsAddress, accAddr sdk.AccAddress) {
+	vac.distributionAddrCache[string(consAddr)] = accAddr
+}
+
+func (vac *ValAddrCache) GetDistributionAddr(ctx sdk.Context, consAddr sdk.ConsAddress) sdk.AccAddress {
+	if value, ok := vac.distributionAddrCache[string(consAddr)]; ok {
+		return value
+	}
+	validator, found := vac.stakeKeeper.GetValidatorByConsAddr(ctx, consAddr)
+	if !found {
+		panic(fmt.Errorf("can't load validator with consensus address %s", consAddr.String()))
+	}
+	distributionAddr := validator.DistributionAddr
+	vac.SetDistributionAddr(consAddr, distributionAddr)
+	return distributionAddr
+}
+
+func distributeFeeBEPHHH(ctx sdk.Context, am auth.AccountKeeper, valAddrCache *ValAddrCache, publishBlockFee bool, stakeKeeper stake.Keeper) (blockFee pub.BlockFee) {
+	// TODO: needs to add BSC ratio Fee
+	fee := fees.Pool.BlockFees()
+	blockFee = pub.BlockFee{Height: ctx.BlockHeader().Height}
+	if fee.IsEmpty() {
+		// no fees in this block
+		return
+	}
+
+	proposerValAddr := ctx.BlockHeader().ProposerAddress
+	proposerDistributionAddr := valAddrCache.GetDistributionAddr(ctx, proposerValAddr)
+
+	// distrubute proposer rewards
+	proposerRewards := sdk.Coins{}
+	feeForAllRewards := sdk.Coins{}
+	var baseProposerRewardRatio int64 = 1  // 1%
+	var bonusProposerRewardRatio int64 = 4 // 4%
+	voteNum := int64(len(ctx.VoteInfos()))
+	// TODO: ensure it's the right way to get current validators
```

Comment:
I think it is not the right way to get it. We need to design the validator storage layout.